internal/runner/runner.go

```diff
@@ -192,15 +232,76 @@ func (r *Runner) Id() string {
 // Start starts the runner by registering the runner with the Waypoint
 // server. This will spawn goroutines for management. This will return after
 // registration so this should not be executed in a goroutine.
-func (r *Runner) Start() error {
+func (r *Runner) Start(ctx context.Context) error {
 	if r.shutdown {
 		return ErrClosed
 	}
 
 	log := r.logger
 
+	// We set this to true if we're going through the adoption process.
+	adopt := false
+	tokenCtx := ctx
+
+	// Check if we have a token in our state directory. If we do, then put
+	// it on our context (so we use it) and mark that we want to adopt. It
+	// may be counterintuitive that we want to adopt if we have a token,
+	// but we use that as a way to verify the token and in the future do rotation.
+	if t, err := r.stateGetToken(); err != nil {
+		return err
+	} else if t != "" {
+		adopt = true
+		log.Debug("will use prior token from state directory")
+		tokenCtx = serverclient.TokenWithContext(tokenCtx, t)
+		r.runningCtx = serverclient.TokenWithContext(r.runningCtx, t)
+	}
+
+	// If we have a cookie set, we always adopt.
+	if r.cookie != "" {
+		adopt = true
+		tokenCtx = metadata.NewOutgoingContext(
+			tokenCtx,
+			metadata.New(map[string]string{
+				"cookie": r.cookie,
+			}),
+		)
+	} else if !adopt {
+		log.Warn("cookie not set for runner, will skip adoption process")
+	}
+
+	if adopt {
+		// Register and initialize the adoption flow (if necessary) by requesting
+		// our token.
+		log.Debug("requesting token with RunnerToken (initiates adoption)")
+		tokenResp, err := r.client.RunnerToken(tokenCtx, &pb.RunnerTokenRequest{
+			Runner: r.runner,
+		})
```

Comment:
Just so that I'm clear, this becomes a sort of "door knock" protocol if the runner already has a token. Ie, the server will say "hey, yup, that's a good token!" and just return the same token.